File: fraiseql-clojure/test/fraiseql/export_types_test.clj
(ns fraiseql.export-types-test
  (:require [clojure.test :refer [deftest is testing before after]]
            [fraiseql.schema :as schema]
            [cheshire.core :as json]
            [clojure.java.io :as io]))

(defn setup [] (schema/reset))
(defn teardown [] (schema/reset))

(deftest export-types-minimal-single-type
  (testing "should export minimal schema with single type"
    (setup)
    (schema/register-type "User"
      {:id {:type "ID" :nullable false}
       :name {:type "String" :nullable false}
       :email {:type "String" :nullable false}}
      "User in the system")

    (let [json (schema/export-types true)
          parsed (json/parse-string json)]

      (is (contains? parsed "types"))
      (is (vector? (get parsed "types")))
      (is (= 1 (count (get parsed "types"))))
      (is (not (contains? parsed "queries")))
      (is (not (contains? parsed "mutations")))
      (is (not (contains? parsed "observers")))
      (is (not (contains? parsed "authz_policies")))

      (let [user-def (first (get parsed "types"))]
        (is (= "User" (get user-def "name")))
        (is (= "User in the system" (get user-def "description")))))
    (teardown)))

(deftest export-types-multiple-types
  (testing "should export minimal schema with multiple types"
    (setup)
    (schema/register-type "User"
      {:id {:type "ID" :nullable false}
       :name {:type "String" :nullable false}})

    (schema/register-type "Post"
      {:id {:type "ID" :nullable false}
       :title {:type "String" :nullable false}
       :authorId {:type "ID" :nullable false}})

    (let [json (schema/export-types true)
          parsed (json/parse-string json)
          types (get parsed "types")]

      (is (= 2 (count types)))

      (let [type-names (map #(get % "name") types)]
        (is (some #(= "User" %) type-names))
        (is (some #(= "Post" %) type-names))))
    (teardown)))

(deftest export-types-no-queries
  (testing "should not include queries in minimal export"
    (setup)
    (schema/register-type "User"
      {:id {:type "ID" :nullable false}})

    (let [json (schema/export-types true)
          parsed (json/parse-string json)]

      (is (contains? parsed "types"))
      (is (not (contains? parsed "queries")))
      (is (not (contains? parsed "mutations"))))
    (teardown)))

(deftest export-types-compact-format
  (testing "should export compact format when pretty is false"
    (setup)
    (schema/register-type "User"
      {:id {:type "ID" :nullable false}})

    (let [compact (schema/export-types false)
          pretty (schema/export-types true)
          parsed (json/parse-string compact)]

      (is (contains? parsed "types"))
      (is (<= (count compact) (count pretty))))
    (teardown)))

(deftest export-types-pretty-format
  (testing "should export pretty format when pretty is true"
    (setup)
    (schema/register-type "User"
      {:id {:type "ID" :nullable false}})

    (let [json (schema/export-types true)]
      (is (clojure.string/includes? json "\n")))
    (teardown)))

(deftest export-types-to-file
  (testing "should export types to file"
    (setup)
    (schema/register-type "User"
      {:id {:type "ID" :nullable false}
       :name {:type "String" :nullable false}})

    (let [tmp-file "/tmp/fraiseql_types_test_clojure.json"]
      (when (.exists (io/file tmp-file))
        (io/delete-file tmp-file))

      (schema/export-types-file tmp-file)

      (is (.exists (io/file tmp-file)))

      (let [content (slurp tmp-file)
            parsed (json/parse-string content)]
        (is (contains? parsed "types"))
        (is (= 1 (count (get parsed "types")))))

      (io/delete-file tmp-file))
    (teardown)))

(deftest export-types-empty
  (testing "should handle empty schema gracefully"
    (setup)

    (let [json (schema/export-types true)
          parsed (json/parse-string json)]

      (is (contains? parsed "types"))
      (is (vector? (get parsed "types")))
      (is (= 0 (count (get parsed "types")))))
    (teardown)))

File: fraiseql-nodejs/tests/export-types.test.ts
import { describe, it, expect, beforeEach, afterEach } from '@jest/globals';
import { Schema } from '../src/schema';
import * as fs from 'fs';
import * as path from 'path';

/**
 * Tests for minimal types.json export (TOML-based workflow)
 *
 * Validates that exportTypes() function generates minimal schema
 * with only types (no queries, mutations, observers, security, etc.)
 */
describe('Export Types - Minimal Schema Export', () => {
  beforeEach(() => {
    // Reset schema registry before each test
    Schema.reset();
  });

  afterEach(() => {
    // Clean up after tests
    Schema.reset();
  });

  it('should export minimal schema with single type', () => {
    // Register a single type
    Schema.registerType('User', {
      fields: {
        id: { type: 'ID', nullable: false },
        name: { type: 'String', nullable: false },
        email: { type: 'String', nullable: false },
      },
      description: 'User in the system',
    });

    // Export minimal types
    const json = Schema.exportTypes(true);
    const parsed = JSON.parse(json);

    // Should have types section
    expect(parsed).toHaveProperty('types');
    expect(Array.isArray(parsed.types)).toBe(true);
    expect(parsed.types).toHaveLength(1);

    // Should NOT have queries, mutations, observers
    expect(parsed).not.toHaveProperty('queries');
    expect(parsed).not.toHaveProperty('mutations');
    expect(parsed).not.toHaveProperty('observers');
    expect(parsed).not.toHaveProperty('authz_policies');

    // Verify User type
    const userDef = parsed.types[0];
    expect(userDef.name).toBe('User');
    expect(userDef.description).toBe('User in the system');
  });

  it('should export minimal schema with multiple types', () => {
    // Register User type
    Schema.registerType('User', {
      fields: {
        id: { type: 'ID', nullable: false },
        name: { type: 'String', nullable: false },
      },
    });

    // Register Post type
    Schema.registerType('Post', {
      fields: {
        id: { type: 'ID', nullable: false },
        title: { type: 'String', nullable: false },
        authorId: { type: 'ID', nullable: false },
      },
    });

    // Export minimal
    const json = Schema.exportTypes(true);
    const parsed = JSON.parse(json);

    // Check types count
    expect(parsed.types).toHaveLength(2);

    // Verify both types present
    const typeNames = parsed.types.map((t: any) => t.name);
    expect(typeNames).toContain('User');
    expect(typeNames).toContain('Post');
  });

  it('should not include queries in minimal export', () => {
    // Register type
    Schema.registerType('User', {
      fields: {
        id: { type: 'ID', nullable: false },
      },
    });

    // Export minimal
    const json = Schema.exportTypes(true);
    const parsed = JSON.parse(json);

    // Should have types
    expect(parsed).toHaveProperty('types');

    // Should NOT have queries
    expect(parsed).not.toHaveProperty('queries');
    expect(parsed).not.toHaveProperty('mutations');
  });

  it('should export compact format when pretty is false', () => {
    Schema.registerType('User', {
      fields: {
        id: { type: 'ID', nullable: false },
      },
    });

    // Export compact
    const compact = Schema.exportTypes(false);

    // Should be valid JSON
    const parsed = JSON.parse(compact);
    expect(parsed).toHaveProperty('types');

    // Compact JSON should be smaller than pretty-printed
    const pretty = Schema.exportTypes(true);
    expect(compact.length).toBeLessThan(pretty.length);
  });

  it('should export pretty format when pretty is true', () => {
    Schema.registerType('User', {
      fields: {
        id: { type: 'ID', nullable: false },
      },
    });

    // Export pretty
    const json = Schema.exportTypes(true);

    // Should contain newlines (pretty format)
    expect(json).toContain('\n');

    // Should be valid JSON
    const parsed = JSON.parse(json);
    expect(parsed).toHaveProperty('types');
  });

  it('should export types to file', () => {
    Schema.registerType('User', {
      fields: {
        id: { type: 'ID', nullable: false },
        name: { type: 'String', nullable: false },
      },
    });

    // Export to temporary file
    const tmpFile = '/tmp/fraiseql_types_test_nodejs.json';

    // Remove file if exists
    if (fs.existsSync(tmpFile)) {
      fs.unlinkSync(tmpFile);
    }

    // Export to file
    Schema.exportTypesFile(tmpFile);

    // Verify file exists and is valid JSON
    expect(fs.existsSync(tmpFile)).toBe(true);

    const content = fs.readFileSync(tmpFile, 'utf-8');
    const parsed = JSON.parse(content);

    expect(parsed).toHaveProperty('types');
    expect(parsed.types).toHaveLength(1);

    // Cleanup
    fs.unlinkSync(tmpFile);
  });

  it('should handle empty schema gracefully', () => {
    // Export with no types registered
    const json = Schema.exportTypes(true);
    const parsed = JSON.parse(json);

    // Should still have types key (as empty array)
    expect(parsed).toHaveProperty('types');
    expect(Array.isArray(parsed.types)).toBe(true);
    expect(parsed.types).toHaveLength(0);
  });
});
